src/gax/src/paginator.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use futures::stream::unfold;
+use futures::{Stream, StreamExt};
+use pin_project::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+
+/// Describes a type that can be iterated over asyncly when used with [Paginator].
+pub trait PageableResponse {
+    fn next_page_token(&self) -> String;
+}
+
+/// An adapter that converts list RPCs as defined by [AIP-4233](https://google.aip.dev/client-libraries/4233)
+/// into a [futures::Stream] that can be iterated over in an async fashion.
+#[pin_project]
+pub struct Paginator<T, E> {
+    #[pin]
+    stream: Pin<Box<dyn Stream<Item = Result<T, E>>>>,
+}
+
+impl<T, E> Paginator<T, E>
+where
+    T: PageableResponse,
+{
+    /// Creates a new [Paginator] given the initial page token and a function
+    /// to fetch the next [PageableResponse].
+    pub fn new<F>(seed_token: String, execute: impl Fn(String) -> F + Clone + 'static) -> Self
+    where
+        // TODO(codyoss): This, and other types, may also need send here in practice.
+        F: Future<Output = Result<T, E>> + 'static,
+    {
+        let stream = unfold(IterState::Token(seed_token), move |state| {
+            let execute = execute.clone();
+            async move {
+                let token = match state {
+                    IterState::Token(token) => token,
+                    IterState::End => return None,
+                };
+                let resp = match execute(token).await {
+                    Ok(page_resp) => {
+                        let tok = page_resp.next_page_token();
+                        let next_state = if tok.is_empty() {
+                            IterState::End
+                        } else {
+                            IterState::Token(tok)
+                        };
+                        Some((Ok(page_resp), next_state))
+                    }
+                    Err(e) => Some((Err(e), IterState::End)),
+                };
+                resp
+            }
+        });
+        Self {
+            stream: Box::pin(stream),
+        }
+    }
+
+    /// Returns the next mutation of the wrapped stream.
+    pub fn next(&mut self) -> futures::stream::Next<'_, Self> {
+        StreamExt::next(self)
+    }
+}
+
+enum IterState {
```

Comment:
This looks like a [std::ops::ControlFlow<String,()>](https://doc.rust-lang.org/beta/std/ops/enum.ControlFlow.html) Should we use that instead?